h2o-algos/src/main/java/hex/gram/Gram.java

```diff
@@ -667,6 +669,28 @@ public final void addRow(DataInfo.Row row, double w) {
     else
       addRowSparse(row, w);
   }
+
+  public final void addElementIJ(DataInfo.Row rowi, double wi, DataInfo.Row rowj, double wj, int ithRow, int jthRow,
+                                 int[] catOffsets) {
+    // nums
+    double elementij = 0.0;
+    double weights = wi*wj;
+    for(int i = 0; i < _denseN; ++i)  {
+      elementij += weights*rowi.numVals[i]*rowj.numVals[i]; // multiply numerical parts of columns
+    }
+
+    // cat X cat
+    if (catOffsets != null) {
```

Comment:
Tomas: 

That is a bug in the categorical data treatment on my part.  Thanks!  After I fixed it, the test results are much better.  The eigenvectors/eigenvalues match exactly using original PCA and my side dataset PCA.  Thanks!

I am doing a dot product between rows and will adopt your proposal.  Thanks, again.  Wendy